python/beeai_framework/agents/rag/agent.py

```diff
@@ -0,0 +1,153 @@
+# Copyright 2025 © BeeAI a Series of LF Projects, LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from enum import Enum
+
+from pydantic import BaseModel, InstanceOf
+
+from beeai_framework.agents import AgentMeta, BaseAgent, BaseAgentRunOptions
+from beeai_framework.agents.rag.prompts import QUERY_IMPROVEMENT_PROMPT
+from beeai_framework.backend import AnyMessage, AssistantMessage, ChatModel, SystemMessage, UserMessage
+from beeai_framework.backend.types import DocumentWithScore
+from beeai_framework.backend.vector_store import VectorStore
+from beeai_framework.context import Run, RunContext
+from beeai_framework.emitter import Emitter
+from beeai_framework.memory import BaseMemory
+from beeai_framework.retrieval.document_processors.document_processors import DocumentsRerankWithLLM
+
+
+class State(BaseModel):
+    final_answer: str
+
+
+class RunInput(BaseModel):
+    message: InstanceOf[AnyMessage]
+
+
+class RAGAgentRunOptions(BaseAgentRunOptions):
+    max_retries: int | None = None
+
+
+class RAGAgentRunOutput(BaseModel):
+    message: InstanceOf[AnyMessage]
+
+
+class FallbackStrategy(str, Enum):
+    NONE = "none"
+    INTERNAL_KNOWLEDGE = "internal_knowledge"
+    WEB_SEARCH = "web_search"
+
+
+class RAGAgent(BaseAgent[RAGAgentRunOutput]):
+    memory: BaseMemory | None = None
```

Comment:
wrong implementation of the abstract class, you should set memory as a private propery (`_memory`) and use getters and setters